libsolidity/AST.cpp

```diff
@@ -673,6 +676,106 @@ void Identifier::checkTypeRequirements()
 		BOOST_THROW_EXCEPTION(createTypeError("Declaration referenced before type could be determined."));
 }
 
+void FunctionIdentifier::overloadResolution()
+{
+	solAssert(!m_overloadedDeclarations.empty(), "FunctionIdentifier not resolved.");
+	solAssert(!m_referencedDeclaration, "Referenced declaration should be null before overload resolution.");
+
+	if (m_overloadedDeclarations.size() == 1)
+	{
+		setReferencedDeclaration(**m_overloadedDeclarations.begin());
+		return;
+	}
+
+	bool resolved = false;
+	auto const& arguments = getFunctionCall().getArguments();
+	auto const& argumentNames = getFunctionCall().getNames();
+
+	if (argumentNames.empty())
+	{
+		// positional arguments
+		for (auto&& declaration: m_overloadedDeclarations)
+		{
+			auto type = declaration->getType();
+			FunctionType const& functionType = dynamic_cast<FunctionType const&>(*type);
```

Comment:
I would actually prefer the reversed behaviour:
Since this is a dynamic cast, we actually know its type, so `auto` is good idea to shorten the code.
On the other hand, using things like `auto type` above hides the information whether `type` is a `TypePointer` or a `Type` and thus can lead to serious errors when e.g. comparing types by equality.